File: src/cc/Hyperspace/ConnectionHandler.cc
/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include "Common/Error.h"
#include "Common/Exception.h"
#include "Common/StringExt.h"

#include "AsyncComm/ApplicationQueue.h"

/**
#include "RequestHandlerCreateTable.h"
#include "RequestHandlerGetSchema.h"
#include "RequestHandlerStatus.h"
**/

#include "Protocol.h"
#include "ConnectionHandler.h"


using namespace hypertable;
using namespace Hyperspace;

/**
 *
 */
void ConnectionHandler::handle(EventPtr &eventPtr) {
  short command = -1;

  if (eventPtr->type == Event::MESSAGE) {
    ApplicationHandler *requestHandler = 0;

    //eventPtr->Display()

    try {
      if (eventPtr->messageLen < sizeof(int16_t)) {
	std::string message = "Truncated Request";
	throw new ProtocolException(message);
      }
      memcpy(&command, eventPtr->message, sizeof(int16_t));

    }
    catch (ProtocolException &e) {
      ResponseCallback cb(mComm, eventPtr);
      std::string errMsg = e.what();
      LOG_VA_ERROR("Protocol error '%s'", e.what());
      cb.error(Error::PROTOCOL_ERROR, errMsg);
    }
  }
  else if (eventPtr->type == Event::CONNECTION_ESTABLISHED) {
    LOG_VA_INFO("%s", eventPtr->toString().c_str());    
  }
  else if (eventPtr->type == Event::DISCONNECT) {
    // do something here!!!
    LOG_VA_INFO("%s : Closing all open handles", eventPtr->toString().c_str());
  }
  else {
    LOG_VA_INFO("%s", eventPtr->toString().c_str());
  }

}

File: src/java/org/hypertable/DfsBroker/hadoop/ConnectionHandler.java
/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package org.hypertable.DfsBroker.hadoop;

import java.net.ProtocolException;
import java.util.logging.Logger;
import org.hypertable.AsyncComm.ApplicationHandler;
import org.hypertable.AsyncComm.ApplicationQueue;
import org.hypertable.AsyncComm.Comm;
import org.hypertable.AsyncComm.DispatchHandler;
import org.hypertable.AsyncComm.Event;
import org.hypertable.AsyncComm.ResponseCallback;
import org.hypertable.Common.Error;

/**
 * This is the connection dispatch handler that gets registered with AsyncComm
 * for each incoming connection.  For MESSAGE events, it inspects the incoming
 * message, extracts the command code, creates a RequestHandler to carry out the
 * request and enque's it on the application work queue.  For DISCONNECT events,
 * it purges all of the open file descriptors that were created on the
 * connection.  All other events are logged.
 */
public class ConnectionHandler implements DispatchHandler {

    static final Logger log = Logger.getLogger("org.hypertable.DfsBroker.hadoop");

    public ConnectionHandler(Comm comm, ApplicationQueue appQueue, HdfsBroker broker) {
	mComm = comm;
	mAppQueue = appQueue;
	mBroker = broker;
    }

    public void handle(Event event) {
	short command;

	if (event.type == Event.Type.MESSAGE) {
	    //log.info(event.toString());

	    ApplicationHandler requestHandler;

	    event.msg.buf.position(event.msg.headerLen);
	    command = event.msg.buf.getShort();

	    switch (command) {
	    case Protocol.COMMAND_OPEN:
		requestHandler = new RequestHandlerOpen(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_CLOSE:
		requestHandler = new RequestHandlerClose(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_CREATE:
		requestHandler = new RequestHandlerCreate(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_LENGTH:
		requestHandler = new RequestHandlerLength(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_READ:
		requestHandler = new RequestHandlerRead(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_WRITE:
		requestHandler = new RequestHandlerWrite(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_SEEK:
		requestHandler = new RequestHandlerSeek(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_REMOVE:
		requestHandler = new RequestHandlerRemove(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_PREAD:
		requestHandler = new RequestHandlerPositionRead(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_MKDIRS:
		requestHandler = new RequestHandlerMkdirs(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_SHUTDOWN:
		requestHandler = new RequestHandlerShutdown(mComm, mAppQueue, event);
		break;
	    case Protocol.COMMAND_STATUS:
		requestHandler = new RequestHandlerStatus(mComm, mAppQueue, event);
		break;
	    case Protocol.COMMAND_FLUSH:
		requestHandler = new RequestHandlerFlush(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_RMDIR:
		requestHandler = new RequestHandlerRmdir(mComm, mBroker, event);
		break;
	    case Protocol.COMMAND_READDIR:
		requestHandler = new RequestHandlerReaddir(mComm, mBroker, event);
		break;
	    default:
		ResponseCallback cb = new ResponseCallback(mComm, event);
		log.severe("Command code " + command + " not implemented");
		cb.error(Error.PROTOCOL_ERROR, "Command code " + command + " not implemented");
		return;
	    }

	    mAppQueue.Add(requestHandler);

	}
	else if (event.type == Event.Type.DISCONNECT) {
	    log.info(event.toString() + " : Closing all open handles from " + event.addr);
	    OpenFileMap ofMap = mBroker.GetOpenFileMap();
	    ofMap.RemoveAll(event.addr);
	}
	else
	    log.info(event.toString());
    }

    private Comm mComm;
    private ApplicationQueue mAppQueue;
    private HdfsBroker mBroker;
}

